UNI-165 python build

marionleborgne opened the pull request.

marionleborgne commented:
@lscheinkman

oxtopus commented on unicorn/backend/unicorn_backend/model_runner.py:
```diff
@@ -22,15 +22,19 @@
 
+# Update pyproj datadir to point to the frozen directory
+# See http://cx-freeze.readthedocs.org/en/latest/faq.html#using-data-files
+if getattr(sys, 'frozen', False):
```
Need single quotes per [nupic python style guide](https://github.com/numenta/nupic/wiki/Python-Style-Guide#quotes)

oxtopus commented on unicorn/backend/unicorn_backend/package_model_runner/freeze_model_runner.py:
```diff
@@ -0,0 +1,18 @@
+                 '/algorithms/modelSelection/anomaly_params_random_encoder/paramOrder.csv',
+                 'htmengine/algorithms/modelSelection/anomaly_params_random_encoder/paramOrder.csv')]
+
+executables = [cx_Freeze.Executable('model_runner.py', targetName = 'mr')]
```
`targetName = 'mr'` -> `targetName="mr"`

oxtopus commented:
Consider incorporating, as there is some overlap:

* https://github.com/numenta/nupic/pull/2617
* https://github.com/numenta/numenta-apps/pull/353
* https://github.com/numenta/numenta-apps/pull/352

shantanoo commented on unicorn/backend/unicorn_backend/package_model_runner/freeze_model_runner.py:
```diff
@@ -0,0 +1,21 @@
+import cx_Freeze
```
Missing copyright notice?

marionleborgne replied:
Thanks! I added it.

marionleborgne commented:
@lscheinkman it's ready :grinning:

lscheinkman commented on htmengine/requirements.txt:
```diff
@@ -4,6 +4,6 @@ sqlalchemy==0.9.4
 validictory==0.9.1
 mysql-python==1.2.5
-nupic==0.3.2
+nupic==0.3.3
```
@marionleborgne The required nupic version is [0.3.4](https://github.com/numenta/nupic/releases/tag/0.3.4)

vitaly-krugl replied:
@marionleborgne: Please don't change htmengine's requirements.txt in this unicorn PR. Unicorn is no longer using htmengine and this nupic version change is properly addressed in PR #352.

CC @lscheinkman

marionleborgne replied:
@vitaly-krugl Ok. I am leaving nupic to version 0.3.4. Can you confirm it is the correct one?

lscheinkman commented on unicorn/backend/unicorn_backend/package_model_runner/freeze_model_runner.py:
```diff
@@ -0,0 +1,60 @@
+
+
+
+zipIncludes = generate_zip_includes("%s/backend/unicorn_backend/nupic/nupic" 
```
@marionleborgne Are you sure nupic is always installed on this location?

lscheinkman replied:
Use  [os.path.join](https://docs.python.org/2/library/os.path.html#os.path.join) instead of manually generating path names

marionleborgne replied:
updated.

lscheinkman commented on unicorn/backend/unicorn_backend/package_model_runner/package.sh:
```diff
@@ -0,0 +1,17 @@
+rm -rf build/ dist/ nupic/
+
+# install nupic locally
+pip install --target $BASE_DIR/unicorn/backend/unicorn_backend/nupic nupic
```
You should use the version specified in the `requirements.txt`

marionleborgne replied:
Ok done. FYI, I removed this script and consolidated it with `freeze_model_runner.py`.

lscheinkman commented on unicorn/backend/unicorn_backend/package_model_runner/package.sh:
```diff
@@ -0,0 +1,17 @@
+#!/usr/bin/env bash
```
we should consolidate the `bash` and `python` portions of this process into a single file. It is hard to maintain when part of the logic is in `package.sh` and the rest in `freeze_model_runner.py`

marionleborgne replied:
done. thanks.

lscheinkman commented on unicorn/frontend/lib/ModelServer.js:
```diff
@@ -27,7 +27,7 @@ import path from 'path';
 
 const MODEL_RUNNER_PATH = path.join(
-  __dirname, '..', '..', 'backend', 'unicorn_backend', 'model_runner.py'
+  __dirname, '..', '..', 'backend', 'unicorn_backend', 'dist', 'mr'
```
`model_runner` is more descriptive than `mr`

marionleborgne replied:
changed to model_runner

lscheinkman commented on unicorn/package.json:
```diff
@@ -16,16 +16,17 @@
     "electron": "electron .",
     "electron:debug": "electron --debug=5858 .",
     "electron:packager": "electron-packager . Unicorn --arch=x64 --icon=assets/unicorn.icns --ignore=.git --ignore=node_modules/electron-* --overwrite --platform=darwin --prune --version=0.32.3",
+    "prepare:python" : "./backend/unicorn_backend/package_model_runner/package.sh",
```
@marionleborgne Unicorn is using [electron-prebuilt v0.33.3](https://github.com/numenta/numenta-apps/blob/master/unicorn/package.json#L95) and `electron:packager` is using `--version=0.32.3`. The `electron` versions should match.

lscheinkman commented on unicorn/package.json:
```diff
@@ -16,16 +16,17 @@
     "clean:npm": "rm -Rf node_modules/",
     "clean:db": "rm -Rf frontend/database/data/*.json",
-    "desktop": "UNICORN_TARGET=desktop npm run prepare && npm run electron",
+    "desktop": "UNICORN_TARGET=desktop npm run prepare && npm run prepare:python && npm run electron",
```
@marionleborgne Are you sure we need to run `cx_freeze` every time we launch the `desktop` app?

marionleborgne replied:
@lscheinkman I moved it from the `desktop` option to the `prepare` one. That way cx_freeze won't be launched every time we run the desktop app. Does that work for you too?

lscheinkman commented on htmengine/requirements.txt:
```diff
@@ -4,6 +4,6 @@ sqlalchemy==0.9.4
 msgpack-python==0.3.0
 validictory==0.9.1
 mysql-python==1.2.5
```
Why `model_runner` requires `mysql-python`?

vitaly-krugl replied:
@lscheinkman: this is not the unicorn model_runner's requirements.txt - it's htmengine's. See my other comment about this next to the nupic version. In htmengine, we configure sqlalchemy to use mysql-python as the underlying driver, so we have to specify a dependency on it. sqlalchemy package iteself doesn't have that dependency.

marionleborgne replied:
@lscheinkman @vitaly-krugl is there something that remains to be addressed here? can I leave the `requirements.txt` of the htmengine like that, since it's not used by the model_runner anymore?

lscheinkman commented on unicorn/package.json:
```diff
@@ -16,16 +16,17 @@
     "desktop:debug": "UNICORN_TARGET=desktop npm run prepare && npm run electron:debug",
     "electron": "electron .",
     "electron:debug": "electron --debug=5858 .",
     "electron:packager": "electron-packager . Unicorn --arch=x64 --icon=assets/unicorn.icns --ignore=.git --ignore=node_modules/electron-* --overwrite --platform=darwin --prune --version=0.32.3",
```
@marionleborgne Unicorn is using [electron-prebuilt v0.33.3](https://github.com/numenta/numenta-apps/blob/master/unicorn/package.json#L95) and `electron:packager` is using `--version=0.32.3`. The `electron` versions should match.

marionleborgne replied:
updated. thanks.

lscheinkman commented on unicorn/backend/unicorn_backend/package_model_runner/package.sh:
```diff
@@ -0,0 +1,17 @@
+#!/usr/bin/env bash
+
+BASE_DIR=/opt/numenta/products/numenta-apps
```
Why do you need this hardcoded absolute path?

marionleborgne replied:
This script has been consolidated in `freeze_model_runner.py` and there is no absolute path anymore. this was indeed unnecessary.

lscheinkman commented on unicorn/backend/unicorn_backend/package_model_runner/freeze_model_runner.py:
```diff
@@ -0,0 +1,60 @@
+import pyproj
+import os
+
+BASE_DIR = "/opt/numenta/products/numenta-apps/unicorn"
```
Why do you need this hardcoded absolute path?

marionleborgne replied:
Removed.

lscheinkman commented on unicorn/backend/unicorn_backend/package_model_runner/freeze_model_runner.py:
```diff
@@ -0,0 +1,60 @@
+
+
+
+def generate_zip_includes(base_path, directory_name):
```
You should avoid using native path separators like `/` . 
Try using [os.sep](https://docs.python.org/2/library/os.html#os.sep) or [os.path](https://docs.python.org/2/library/os.path.html#module-os.path) to parse and manipulate file system path in a cross platform way.

lscheinkman commented:
@unixorn Can you review this logic as well to make sure it is compatible with Mac `jenkins` node and the infrastructure scripts?

unixorn commented:
@lscheinkman it looks ok to me overall, once your comments are addressed.

vitaly-krugl commented:
Please remove changes to htmengine/requirements.txt from this PR.

marionleborgne commented:
@lscheinkman: I think I addressed all your comments. Thanks for the feedback.

@vitaly-krugl: I noticed that the file `stats_schema.json` was not found. So I had to update the model runner to tell it where to find the `stats_schema.json` file depending on wether or not the `model_runner` is running in its frozen version (meaning packaged with `cx_freeze`). See  https://github.com/numenta/numenta-apps/pull/354/files#diff-2723f200c1544575e0a7f3a38c932ec9R137 and related stackoverflow answer: http://stackoverflow.com/a/2632297

cc @unixorn

lscheinkman commented on unicorn/backend/unicorn_backend/freeze_model_runner.py:
```diff
@@ -0,0 +1,93 @@
+  :param base_path: path to the parent dir of the package to include in the zip.
+  :param directory_name: target directory name.
+  """
+  skip_count = len(base_path.split("/"))
```
use `os.sep` instead of `'/'` here and everywhere else in this function

marionleborgne replied:
Sorry, I missed that one. Updated, thanks.

lscheinkman commented on unicorn/backend/unicorn_backend/freeze_model_runner.py:
```diff
@@ -0,0 +1,93 @@
+  # Install nupic locally and include it in the library.zip generated by 
+  # cx_freeze.
+  modelRunnerDir = os.path.dirname(os.path.realpath(__file__))
+  nupicInstall = ("pip install --target %s nupic==0.3.4"
```
Nupic version number should come from the `requirements.txt. It should not be hardcoded in the script

marionleborgne replied:
Updated. I am using the nupic version from `requirements.txt`

lscheinkman commented on unicorn/backend/unicorn_backend/freeze_model_runner.py:
```diff
@@ -0,0 +1,93 @@
+  same directory as model_runner.py.
+  """
+  # Initial cleanup.
+  os.system("rm -rf build dist nupic")
```
Should run the clean command relative to `modelRunnerDir` instead of  the current directory

marionleborgne replied:
thanks for the catch! updated

lscheinkman commented on unicorn/package.json:
```diff
@@ -28,9 +28,10 @@
+    "prepare:python" : "python backend/unicorn_backend/freeze_model_runner.py",
     "node-inspector": "node-inspector",
-    "prepare": "npm run prepare:lint && npm run prepare:webpack",
+    "prepare": "npm run prepare:lint && npm run prepare:webpack && npm run prepare:python",
```
Are you sure you want to execute `freeze_model_runner.py` here?
`npm run prepare` is executed every time you start the application with  `npm run desktop`. See [line 27](https://github.com/numenta/numenta-apps/pull/354/files#diff-2b77fd4fad142ba758d6983fb1205f11L27)

marionleborgne replied:
I ended up taking it out of the `prepare` option, so that it is not run as part of `npm run prepare` and `npm run dektop`. I added `npm run prepare:python` in the `build` option.

marionleborgne commented:
@lscheinkman please review my last changes

lscheinkman commented on unicorn/frontend/lib/ModelServer.js:
```diff
@@ -28,7 +28,7 @@ import UserError from './UserError';
 
 const MODEL_RUNNER_PATH = path.join(
-  __dirname, '..', '..', 'backend', 'unicorn_backend', 'model_runner.py'
+  __dirname, '..', '..', 'dist', 'model_runner'
```
@marionleborgne This will only work when the script is frozen (`prepare:python`) and fails when running directly from source `npm run desktop` or `npm run test`. 
Now that  `prepare:python` was removed from `npm run desktop` the application will fail to find `dist/model_runner` if it was never `frozen`. 
For now we should add a note to the `README` file requiring the user to run `npm run prepare:python` before start developing and every time `model_runner.py` or `nupic` is updated.

marionleborgne replied:
I updated the instructions.

lscheinkman commented:
:+1: LGTM 
Just a minor update to the build instructions but the code looks good.
Great Job!

marionleborgne commented:
Thanks @lscheinkman. Merging :) Unicorn phase 1 = completed!
